File: app/src/main/java/com/fox/algorithm/数据结构与算法/leetcode/栈_队列/_20有效的括号.java
package com.fox.algorithm.数据结构与算法.leetcode.栈_队列;

import java.util.HashMap;
import java.util.LinkedList;

/**
 * @Author Fox
 * @Date 2021/3/28 12:33
 */

/*https://leetcode-cn.com/problems/valid-parentheses/solution/you-xiao-de-gua-hao-by-leetcode-solution/
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。
有效字符串需满足：
左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。
输入：s = "()"
输出：true
 */
class _20有效的括号 {

    public boolean isValid(String s) {
        //注意这里的key value 是反过来的
        HashMap<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');

        LinkedList<Character> stack = new LinkedList<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (map.containsKey(ch)) {
                if (stack.isEmpty() || stack.peek() != map.get(ch)) {
                    return false;
                }
                stack.pop();
            } else {
                stack.push(ch);
            }
        }
        return stack.isEmpty();
    }
}

File: app/src/main/java/com/fox/algorithm/数据结构与算法/leetcode/栈_队列/20有效的括号.kt
package com.fox.algorithm.数据结构与算法.leetcode.栈_队列

import java.util.*

/**
 * @Author Fox
 * @Date 2021/3/18 11:03
 */
/*https://leetcode-cn.com/problems/valid-parentheses/solution/you-xiao-de-gua-hao-by-leetcode-solution/
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。

有效字符串需满足：

左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。
输入：s = "()"
输出：true
 */
fun isValid(s: String): Boolean {
    if (s.length % 2 == 1) return false
    //注意这里的key value 是反过来的
    val map = hashMapOf(']' to '[', ')' to '(', '}' to '{')
    val stack = LinkedList<Char>()

    for (i in s.indices) {
        val ch = s[i]
        if (map.containsKey(ch)) {
            if (stack.isEmpty() || stack.peek() != map[ch]) {
                return false
            }
            stack.pop()
        } else {
            stack.push(ch)
        }
    }
    return stack.isEmpty()
}